tests/tas-elasticsearch/src/test/java/org/alfresco/elasticsearch/reindexing/NodesSecondaryAncestorIndexingTests.java

```diff
@@ -0,0 +1,337 @@
+    public void testSecondaryAncestorWithNodeHavingTwoSecondaryChildren()
+    {
+        // then
+        STEP("Verify that searching by ANCESTOR and folderC will find nodes: .");
+        SearchRequest queryAncestorC = req("ANCESTOR:" + folders(L).getNodeRef());
+        searchQueryService.expectResultsFromQuery(queryAncestorC, testUser,
+            // primary descendant
+            folders(M).getName(),
+            // secondary descendants
+            folders(C).getName(),
+            folders(Y).getName(),
+            folders(Z).getName());
+    }
+
+    @Test(groups = TestGroup.SEARCH)
+    public void testSecondaryAncestorWithDocumentAsSecondaryChild()
+    {
+        // then
+        STEP("Verify that searching by ANCESTOR and folderA will find nodes: .");
+        SearchRequest query = req("ANCESTOR:" + folders(A).getNodeRef());
+        searchQueryService.expectResultsFromQuery(query, testUser,
+            // primary descendants
+            folders(B).getName(),
+            folders(C).getName(),
+            // secondary descendant
+            fileInP.getName());
+    }
+
+    @Test(groups = TestGroup.SEARCH)
+    public void testSecondaryAncestorWithNodeHavingComplexSecondaryRelationship()
+    {
+        // then
+        STEP("Verify that all descendant of folderX can be found.");
+        SearchRequest query = req("ANCESTOR:" + folders(X).getNodeRef());
+        searchQueryService.expectResultsFromQuery(query, testUser,
+            // primary descendants
+            folders(Y).getName(),
+            folders(Z).getName(),
+            // secondary descendants
+            folders(B).getName(),
+            folders(C).getName(),
+            folders(K).getName(),
+            folders(L).getName(),
+            folders(M).getName()
+        );
+    }
+
+    /**
+     * Verify that removing secondary parent-child relationship will result in updating ES index: ANCESTOR.
+     * Test changes bellow folders hierarchy:
+     * <pre>
+     * DL
+     *  += fQ
+     *     +
+     *     |
+     *  += fR
+     * </pre>
+     * into:
+     * <pre>
+     * DL
+     *  += fQ
+     *  += fR
+     * </pre>
+     */
+    @Test(groups = TestGroup.SEARCH)
+    public void testSecondaryAncestorWithDeletedSecondaryRelationship()
+    {
+        // given
+        STEP("Add to folderQ a secondary child folderR and verify if it can be found using ANCESTOR index and secondary child node reference.");
+        folders(Q).addSecondaryChild(folders(R));
+
+        STEP("Verify that searching by ANCESTOR and folderQ will find secondary descendant node: folderR.");
+        SearchRequest query = req("ANCESTOR:" + folders(Q).getNodeRef());
+        searchQueryService.expectResultsFromQuery(query, testUser,
+            // secondary descendant
+            folders(R).getName());
+
+        // when
+        STEP("Delete the secondary parent-child relationship between folderQ and FolderR.");
+        folders(Q).removeSecondaryChild(folders(R));
```

Comment:
I think there's a risk that this test will be intermittent. I think we've seen the worst ES offenders are tests which change paths and then expect updates to happen soon afterwards. Could we instead change this to perform the add and remove in the precondition?

Alternatively we could just keep an eye out to see whether this test actually does fail frequently or not. Perhaps it will be fine.

Reply:
I understand your point, but before PTO I'm not able to fix all the tests, which change paths and then expect updates to happen soon afterwards. Lets keep eye on them and I can fix them after my PTO.